{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Relation Data EDA\n",
    "\n",
    "This notebook computes summary statistics and performs sanity checks on the relation data processing pipeline (raw_data -> relations database and sentences tagged with key terms)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use device: cpu\n",
      "---\n",
      "Loading: tokenize\n",
      "With settings: \n",
      "{'model_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt_tokenizer.pt', 'lang': 'en', 'shorthand': 'en_ewt', 'mode': 'predict'}\n",
      "---\n",
      "Loading: pos\n",
      "With settings: \n",
      "{'model_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt_tagger.pt', 'pretrain_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt.pretrain.pt', 'lang': 'en', 'shorthand': 'en_ewt', 'mode': 'predict'}\n",
      "---\n",
      "Loading: lemma\n",
      "With settings: \n",
      "{'model_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt_lemmatizer.pt', 'lang': 'en', 'shorthand': 'en_ewt', 'mode': 'predict'}\n",
      "Building an attentional Seq2Seq model...\n",
      "Using a Bi-LSTM encoder\n",
      "Using soft attention for LSTM.\n",
      "Finetune all embeddings.\n",
      "[Running seq2seq lemmatizer with edit classifier]\n",
      "---\n",
      "Loading: depparse\n",
      "With settings: \n",
      "{'model_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt_parser.pt', 'pretrain_path': '/Users/mattboggess/stanfordnlp_resources/en_ewt_models/en_ewt.pretrain.pt', 'lang': 'en', 'shorthand': 'en_ewt', 'mode': 'predict'}\n",
      "Done loading processors!\n",
      "---\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import json\n",
    "from collections import defaultdict\n",
    "import spacy\n",
    "import stanfordnlp\n",
    "import pandas as pd\n",
    "from spacy_stanfordnlp import StanfordNLPLanguage\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "from ipywidgets import interact\n",
    "\n",
    "# nlp preprocessing pipeline\n",
    "warnings.filterwarnings('ignore')\n",
    "snlp = stanfordnlp.Pipeline(lang=\"en\")\n",
    "nlp = StanfordNLPLanguage(snlp)\n",
    "\n",
    "# fix for importing utils\n",
    "module_path = os.path.abspath(os.path.join('../data_processing'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from data_processing_utils import read_spacy_docs, tag_terms\n",
    "\n",
    "data_dir = '../data/relation_extraction'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What concepts actually match the text?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_sentences = read_spacy_docs(\"../data/preprocessed_data/Biology_2e_sentences_spacy\", nlp)\n",
    "bio_sentences += read_spacy_docs(\"../data/preprocessed_data/Life_Biology_sentences_spacy\", nlp)\n",
    "kb_terms = read_spacy_docs(\"../data/preprocessed_data/Life_Biology_kb_key_terms_spacy\", nlp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tagging sentence 0/44162\n",
      "tagging sentence 500/44162\n",
      "tagging sentence 1000/44162\n",
      "tagging sentence 1500/44162\n",
      "tagging sentence 2000/44162\n",
      "tagging sentence 2500/44162\n",
      "tagging sentence 3000/44162\n",
      "tagging sentence 3500/44162\n",
      "tagging sentence 4000/44162\n",
      "tagging sentence 4500/44162\n",
      "tagging sentence 5000/44162\n",
      "tagging sentence 5500/44162\n",
      "tagging sentence 6000/44162\n",
      "tagging sentence 6500/44162\n",
      "tagging sentence 7000/44162\n",
      "tagging sentence 7500/44162\n",
      "tagging sentence 8000/44162\n",
      "tagging sentence 8500/44162\n",
      "tagging sentence 9000/44162\n",
      "tagging sentence 9500/44162\n",
      "tagging sentence 10000/44162\n",
      "tagging sentence 10500/44162\n",
      "tagging sentence 11000/44162\n",
      "tagging sentence 11500/44162\n",
      "tagging sentence 12000/44162\n",
      "tagging sentence 12500/44162\n",
      "tagging sentence 13000/44162\n",
      "tagging sentence 13500/44162\n",
      "tagging sentence 14000/44162\n",
      "tagging sentence 14500/44162\n",
      "tagging sentence 15000/44162\n",
      "tagging sentence 15500/44162\n",
      "tagging sentence 16000/44162\n",
      "tagging sentence 16500/44162\n",
      "tagging sentence 17000/44162\n",
      "tagging sentence 17500/44162\n",
      "tagging sentence 18000/44162\n",
      "tagging sentence 18500/44162\n",
      "tagging sentence 19000/44162\n",
      "tagging sentence 19500/44162\n",
      "tagging sentence 20000/44162\n",
      "tagging sentence 20500/44162\n",
      "tagging sentence 21000/44162\n",
      "tagging sentence 21500/44162\n",
      "tagging sentence 22000/44162\n",
      "tagging sentence 22500/44162\n",
      "tagging sentence 23000/44162\n",
      "tagging sentence 23500/44162\n",
      "tagging sentence 24000/44162\n",
      "tagging sentence 24500/44162\n",
      "tagging sentence 25000/44162\n",
      "tagging sentence 25500/44162\n",
      "tagging sentence 26000/44162\n",
      "tagging sentence 26500/44162\n",
      "tagging sentence 27000/44162\n",
      "tagging sentence 27500/44162\n",
      "tagging sentence 28000/44162\n",
      "tagging sentence 28500/44162\n",
      "tagging sentence 29000/44162\n",
      "tagging sentence 29500/44162\n",
      "tagging sentence 30000/44162\n",
      "tagging sentence 30500/44162\n",
      "tagging sentence 31000/44162\n",
      "tagging sentence 31500/44162\n",
      "tagging sentence 32000/44162\n",
      "tagging sentence 32500/44162\n",
      "tagging sentence 33000/44162\n",
      "tagging sentence 33500/44162\n",
      "tagging sentence 34000/44162\n",
      "tagging sentence 34500/44162\n",
      "tagging sentence 35000/44162\n",
      "tagging sentence 35500/44162\n",
      "tagging sentence 36000/44162\n",
      "tagging sentence 36500/44162\n",
      "tagging sentence 37000/44162\n",
      "tagging sentence 37500/44162\n",
      "tagging sentence 38000/44162\n",
      "tagging sentence 38500/44162\n",
      "tagging sentence 39000/44162\n",
      "tagging sentence 39500/44162\n",
      "tagging sentence 40000/44162\n",
      "tagging sentence 40500/44162\n",
      "tagging sentence 41000/44162\n",
      "tagging sentence 41500/44162\n",
      "tagging sentence 42000/44162\n",
      "tagging sentence 42500/44162\n",
      "tagging sentence 43000/44162\n",
      "tagging sentence 43500/44162\n",
      "tagging sentence 44000/44162\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "# count occurrences of all terms\n",
    "term_counts = Counter()\n",
    "for i, sentence in enumerate(bio_sentences):\n",
    "    if i % 500 == 0:\n",
    "        print(f\"tagging sentence {i}/{len(bio_sentences)}\")\n",
    "    _, _, term_info = tag_terms(sentence, kb_terms, nlp)\n",
    "    term_counts.update({t: len(term_info[t][\"indices\"]) for t in term_info})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3963/5941 bio kb concepts tagged in text\n"
     ]
    }
   ],
   "source": [
    "with open(\"../data/preprocessed_data/Life_Biology_kb_lexicon.json\", \"r\") as f:\n",
    "    lexicon = json.load(f)\n",
    "\n",
    "tagged_concepts = []\n",
    "non_tagged_concepts = []\n",
    "for concept in lexicon.keys():\n",
    "    if any([lemma in set(term_counts.keys()) for lemma in lexicon[concept][\"lemma_representations\"]]) or \\\n",
    "       any([text in set(term_counts.keys()) for text in lexicon[concept][\"text_representations\"]]):\n",
    "        tagged_concepts.append({concept: lexicon[concept]})\n",
    "    else:\n",
    "        non_tagged_concepts.append({concept: lexicon[concept]})\n",
    "print(f\"{len(tagged_concepts)}/{len(non_tagged_concepts) + len(tagged_concepts)} bio kb concepts tagged in text\")\n",
    "\n",
    "df = {\"concept\": [], \"text\": []}\n",
    "for concept in non_tagged_concepts:\n",
    "    c = list(concept.keys())[0]\n",
    "    df[\"concept\"].append(c)\n",
    "    df[\"text\"].append(concept[c][\"text_representations\"])\n",
    "pd.DataFrame(df).to_excel(\"../data/relation_extraction/diagnostics/not_tagged_concepts.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Relations Database EDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../data/relation_extraction/relations_db.json\", \"r\") as f:\n",
    "    rdb = json.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Statistics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "long_df = {\"relation\": [], \"term_pair\": [], \"count_sentences\": [], \"found_sentence\": []}\n",
    "for relation in rdb:\n",
    "    for term_pair in rdb[relation]: \n",
    "        long_df[\"relation\"].append(relation)\n",
    "        long_df[\"term_pair\"].append(term_pair)\n",
    "        long_df[\"count_sentences\"].append(len(rdb[relation][term_pair][\"sentences\"]))\n",
    "        long_df[\"found_sentence\"].append(len(rdb[relation][term_pair][\"sentences\"]) > 0)\n",
    "long_df = pd.DataFrame(long_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>term_pair</th>\n",
       "      <th colspan=\"2\" halign=\"left\">count_sentences</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>sum</th>\n",
       "      <th>mean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>relation</th>\n",
       "      <th>found_sentence</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">abuts</td>\n",
       "      <td>False</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>26</td>\n",
       "      <td>146</td>\n",
       "      <td>5.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">element</td>\n",
       "      <td>False</td>\n",
       "      <td>443</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>92</td>\n",
       "      <td>629</td>\n",
       "      <td>6.836957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">has-part</td>\n",
       "      <td>False</td>\n",
       "      <td>5704</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>863</td>\n",
       "      <td>6553</td>\n",
       "      <td>7.593279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">has-region</td>\n",
       "      <td>False</td>\n",
       "      <td>2817</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>234</td>\n",
       "      <td>1405</td>\n",
       "      <td>6.004274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">is-at</td>\n",
       "      <td>False</td>\n",
       "      <td>258</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>46</td>\n",
       "      <td>331</td>\n",
       "      <td>7.195652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">is-inside</td>\n",
       "      <td>False</td>\n",
       "      <td>912</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>156</td>\n",
       "      <td>939</td>\n",
       "      <td>6.019231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">is-outside</td>\n",
       "      <td>False</td>\n",
       "      <td>173</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>22</td>\n",
       "      <td>182</td>\n",
       "      <td>8.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>no-relation</td>\n",
       "      <td>True</td>\n",
       "      <td>232471</td>\n",
       "      <td>539374</td>\n",
       "      <td>2.320178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">possesses</td>\n",
       "      <td>False</td>\n",
       "      <td>1787</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>101</td>\n",
       "      <td>408</td>\n",
       "      <td>4.039604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">subclass-of</td>\n",
       "      <td>False</td>\n",
       "      <td>17937</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>1242</td>\n",
       "      <td>3918</td>\n",
       "      <td>3.154589</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           term_pair count_sentences          \n",
       "                               count             sum      mean\n",
       "relation    found_sentence                                    \n",
       "abuts       False                140               0  0.000000\n",
       "            True                  26             146  5.615385\n",
       "element     False                443               0  0.000000\n",
       "            True                  92             629  6.836957\n",
       "has-part    False               5704               0  0.000000\n",
       "            True                 863            6553  7.593279\n",
       "has-region  False               2817               0  0.000000\n",
       "            True                 234            1405  6.004274\n",
       "is-at       False                258               0  0.000000\n",
       "            True                  46             331  7.195652\n",
       "is-inside   False                912               0  0.000000\n",
       "            True                 156             939  6.019231\n",
       "is-outside  False                173               0  0.000000\n",
       "            True                  22             182  8.272727\n",
       "no-relation True              232471          539374  2.320178\n",
       "possesses   False               1787               0  0.000000\n",
       "            True                 101             408  4.039604\n",
       "subclass-of False              17937               0  0.000000\n",
       "            True                1242            3918  3.154589"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = long_df.groupby([\"relation\", \"found_sentence\"]).agg({\"term_pair\": \"count\",\n",
    "                                                                  \"count_sentences\": [\"sum\", \"mean\"]})\n",
    "summary_df.to_csv(\"../data/relation_extraction/summary/relation_counts_summary.csv\", index=False)\n",
    "summary_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sentence Count Distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['no-relation', 'subclass-of', 'has-part', 'possesses', 'has-region', 'is-inside', 'is-at', 'element', 'abuts', 'is-outside'])\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "17152e9d33cb4555ac581b9209b5f980",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='relation', options=('no-relation', 'subclass-of', 'has-part', 'pos…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<function __main__.plot_sentence_counts(relation)>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from ipywidgets import interact\n",
    "\n",
    "def plot_sentence_counts(relation):\n",
    "    \n",
    "    df = long_df.query(f\"count_sentences > 0 & relation == '{relation}'\")\n",
    "    df = df.query(\"count_sentences < 30\")\n",
    "    plt.hist(df.count_sentences, bins=30)\n",
    "    plt.show()\n",
    "    \n",
    "interact(plot_sentence_counts, relation=list(rdb.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multi-Label Term Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "247\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'substance -> object': ('Substance -> Tangible-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'mixture -> substance': ('Mixture -> Substance', ['subclass-of', 'has-part']),\n",
       " 'membrane -> membrane': ('Biomembrane -> Membrane', ['subclass-of', 'is-at']),\n",
       " 'segment of body -> anatomical structure': ('Body-Segment -> Anatomical-Structure',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'bone -> connective tissue': ('Bone -> Connective-Tissue',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'CDNA -> dna': ('CDNA -> DNA', ['subclass-of', 'abuts']),\n",
       " 'circular dna -> dna': ('Circular-DNA -> DNA', ['subclass-of', 'has-part']),\n",
       " 'plasmid -> dna': ('Circular-DNA -> DNA', ['subclass-of', 'has-part']),\n",
       " 'gap junction -> pore': ('Cytoplasmic-Channel -> Pore',\n",
       "  ['subclass-of', 'has-region']),\n",
       " 'diet -> object': ('Diet -> Tangible-Entity', ['subclass-of', 'element']),\n",
       " 'gene -> dna sequence': ('Gene -> DNA-Sequence', ['subclass-of', 'has-part']),\n",
       " 'glycosidic linkage -> polar covalent bond': ('Glycosidic-Linkage -> Polar-Covalent-Bond',\n",
       "  ['subclass-of', 'element']),\n",
       " 'homeodomain -> protein domain': ('Homeodomain -> Protein-Domain',\n",
       "  ['subclass-of', 'has-region']),\n",
       " 'hydrophobic core -> hydrophobic region': ('Hydrophobic-Core -> Hydrophobic-Region',\n",
       "  ['subclass-of', 'abuts']),\n",
       " 'chromosome -> dna': ('Recombinant-Chromosome -> DNA',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'chromosome -> sister chromatid': ('Nonsister-chromatid -> Chromatid',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'chromosome -> chromatid': ('Nonsister-chromatid -> Chromatid',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'nucleic acid probe -> nucleic acid strand': ('Nucleic-Acid-Probe -> Nucleic-Acid-Strand',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'macromolecule -> chemical': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'macromolecule -> chemical entity': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'macromolecule -> substance': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'polymer -> chemical': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'polymer -> chemical entity': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'polymer -> substance': ('Polymer -> Chemical-Entity',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'proteasome -> protein': ('Proteasome -> Protein',\n",
       "  ['subclass-of', 'has-part']),\n",
       " 'seed -> tissue': ('Seedling -> Tissue', ['subclass-of', 'has-part']),\n",
       " 'trimer -> molecule': ('Trimer -> Molecule', ['subclass-of', 'has-part']),\n",
       " 'variable domain -> protein domain': ('Variable-Domain -> Protein-Domain',\n",
       "  ['subclass-of', 'has-region']),\n",
       " 'visible light -> green light': ('ABOX_Visible-Light -> ABOX_Green-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'visible light -> blue light': ('ABOX_Visible-Light -> ABOX_Blue-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'visible light -> yellow light': ('ABOX_Visible-Light -> ABOX_Yellow-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'visible light -> orange light': ('ABOX_Visible-Light -> ABOX_Orange-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'visible light -> red light': ('ABOX_Visible-Light -> ABOX_Red-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'visible light -> violet light': ('ABOX_Visible-Light -> ABOX_Violet-Light',\n",
       "  ['has-part', 'element']),\n",
       " 'mitotic spindle -> microtubule': ('ABOX_Mitotic-Spindle -> ABOX_Microtubule',\n",
       "  ['has-part', 'element']),\n",
       " 'heterozygous -> allele': ('ABOX_Heterozygous -> ABOX_Allele',\n",
       "  ['has-part', 'element']),\n",
       " 'dna sequence -> dna strand': ('ABOX_DNA-Sequence -> ABOX_DNA-Strand',\n",
       "  ['has-part', 'is-at']),\n",
       " 'electron microscope -> electromagnet': ('ABOX_Electron-Microscope -> ABOX_Electromagnet',\n",
       "  ['has-part', 'abuts']),\n",
       " 'prophage -> gene': ('ABOX_Prophage -> ABOX_Gene', ['has-part', 'element']),\n",
       " 'water molecule -> chemical': ('ABOX_Water-Molecule -> Chemical-Entity',\n",
       "  ['has-part', 'abuts']),\n",
       " 'object -> membrane': ('ABOX_Tangible-Entity -> ABOX_Biomembrane',\n",
       "  ['has-part', 'is-inside']),\n",
       " 'atp synthase -> stator': ('ABOX_ATP-Synthase -> ABOX_Stator',\n",
       "  ['has-part', 'element']),\n",
       " 'transcription unit -> terminator sequence': ('ABOX_Transcription-Unit -> ABOX_Terminator-Sequence',\n",
       "  ['has-part', 'abuts']),\n",
       " 'transcription unit -> promoter': ('ABOX_Transcription-Unit -> ABOX_Promoter',\n",
       "  ['has-part', 'abuts']),\n",
       " 'atp synthase -> rod': ('ABOX_ATP-Synthase -> ABOX_Rod',\n",
       "  ['has-part', 'element']),\n",
       " 'atp synthase -> rotor': ('ABOX_ATP-Synthase -> ABOX_Rotor',\n",
       "  ['has-part', 'element']),\n",
       " 'chemical -> water molecule': ('Chemical -> ABOX_Water-Molecule',\n",
       "  ['has-part', 'abuts']),\n",
       " 'gene -> repetitive dna': ('ABOX_Gene -> ABOX_Repetitive-DNA',\n",
       "  ['has-part', 'element']),\n",
       " 'nad plus -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> phosphate group': ('ABOX_ATP -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adenylyl cyclase -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'epidermis -> outer layer': ('ABOX_Epidermis -> ABOX_Outer-Layer',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'water -> surface': ('ABOX_Water -> ABOX_Surface', ['has-region', 'is-at']),\n",
       " 'adenylyl cyclase -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adenylyl cyclase -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'hydrophobic amino acid -> hydrophobic region': ('ABOX_Hydrophobic-Amino-Acid -> ABOX_Hydrophobic-Region',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical entity -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'substance -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'oxygen atom -> phosphate': ('Chemical-Entity -> Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'oxygen atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'hydrogen atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nitrogen atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pgap -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> carbonyl group': ('Molecule -> Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> carbon skeleton': ('Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> carbon skeleton': ('ABOX_Acetyl-CoA -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADH -> carbon skeleton': ('NADH -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> acetyl group': ('Organic-Molecule -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> negatively charge region': ('Organic-Molecule -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'adenylyl cyclase -> amino group': ('Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> amino group': ('Organic-Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adenylyl cyclase -> carboxyl group': ('Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'alcohol dehydrogenase -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pgap -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pyrophosphate -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADPH -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphorus atom -> phosphate': ('Chemical-Entity -> Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphorus atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphate io n -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> amino group': ('Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> amino group': ('Organic-Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> carboxyl group': ('Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADH -> carboxyl group': ('Organic-Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> surface': ('Chemical-Entity -> Surface',\n",
       "  ['has-region', 'is-at']),\n",
       " 'adp -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADH -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical entity -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'substance -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'oxygen atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'hydrogen atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nitrogen atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pgap -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pyrophosphate -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphorus atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphate io n -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> negatively charge region': ('Organic-Molecule -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'nad plus -> negatively charge region': ('Organic-Molecule -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'adenylyl cyclase -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADH -> phosphate group': ('Organic-Molecule -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical entity -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'substance -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'oxygen atom -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'hydrogen atom -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nitrogen atom -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'fadh2 -> acetyl group': ('Organic-Molecule -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pgap -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> acetyl group': ('Organic-Molecule -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> acetyl group': ('Organic-Molecule -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon atom -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pyrophosphate -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> acetyl group': ('Organic-Molecule -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphorus atom -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphate io n -> acetyl group': ('Chemical-Entity -> ABOX_Acetyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'adp -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nad plus -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical entity -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'substance -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'oxygen atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'hydrogen atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'nitrogen atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> carboxyl group': ('Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'pyrophosphate -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphorus atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'phosphate io n -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> carboxyl group': ('Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> amino group': ('Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> amino group': ('Molecule -> ABOX_Amino-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'alcohol dehydrogenase -> carboxyl group': ('Molecule -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atp -> negatively charge region': ('Organic-Molecule -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'acidic amino acid -> protein domain': ('Organic-Molecule -> Protein-Domain',\n",
       "  ['has-region', 'is-at']),\n",
       " 'acidic amino acid -> polypeptide segment': ('Organic-Molecule -> Protein-Domain',\n",
       "  ['has-region', 'is-at']),\n",
       " 'alanine -> protein domain': ('Organic-Molecule -> Protein-Domain',\n",
       "  ['has-region', 'is-at']),\n",
       " 'alanine -> polypeptide segment': ('Organic-Molecule -> Protein-Domain',\n",
       "  ['has-region', 'is-at']),\n",
       " 'alcohol dehydrogenase -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'acetyl coa -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " '3 phosphoglycerate -> carboxyl group': ('ABOX_Three-Phosphoglycerate -> Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " '3 phosphoglycerate -> phosphate group': ('ABOX_Three-Phosphoglycerate -> Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " '3 phosphoglycerate -> carbon skeleton': ('Organic-Molecule -> ABOX_Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> hydroxyl group': ('Molecule -> ABOX_Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'molecule -> negatively charge region': ('Molecule -> ABOX_Negatively-Charged-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'alcohol dehydrogenase -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'chemical -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'chemical entity -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'substance -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'oxygen atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'hydrogen atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'nitrogen atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'pgap -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'carbon atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'pyrophosphate -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'phosphorus atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'phosphate io n -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'atom -> negatively charge region': ('Chemical-Entity -> Hydrophilic-Region',\n",
       "  ['has-region', 'is-at']),\n",
       " 'atom -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADPH -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atom -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'NADPH -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'atom -> carboxyl group': ('Chemical-Entity -> ABOX_Carboxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " '3 phosphoglycerate -> hydroxyl group': ('Organic-Molecule -> Hydroxyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " '3 phosphoglycerate -> carbonyl group': ('Organic-Molecule -> ABOX_Carbonyl-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon dioxide -> phosphate group': ('Chemical-Entity -> ABOX_Phosphate-Group',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'carbon dioxide -> carbon skeleton': ('Chemical-Entity -> Carbon-Skeleton',\n",
       "  ['has-region', 'is-inside']),\n",
       " 'protein enzyme -> er membrane': ('ABOX_Protein-Enzyme -> ABOX_ER-Membrane',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'aqueous solution -> er membrane': ('ABOX_Aqueous-Solution -> ABOX_ER-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'sodium - potassium pump -> phospholipid bilayer': ('ABOX_Sodium-Potassium-Pump -> ABOX_Phospholipid-Bilayer',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'hydrophilic end -> outside face': ('ABOX_Hydrophilic-End -> ABOX_Outside-Face',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'transport membrane protein -> phospholipid bilayer': ('ABOX_Transport-Protein -> ABOX_Phospholipid-Bilayer',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'phospholipid -> phospholipid layer': ('ABOX_Phospholipid -> ABOX_Phospholipid-Layer',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'object -> mitochondrion': ('Spatial-Entity -> ABOX_Mitochondrion',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> central vacuole': ('ABOX_Tangible-Entity -> ABOX_Central-Vacuole',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> aqueous solution': ('ABOX_Hydrophilic-Face -> ABOX_Aqueous-Solution',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> cell-membrane': ('ABOX_Hydrophilic-Face -> ABOX_Plasma-membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> nuclear-envelope': ('ABOX_Hydrophilic-Face -> ABOX_Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> Platyhelminthes': ('ABOX_Spatial-Entity -> ABOX_Platyhelminthes',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> pancrea cell': ('ABOX_Spatial-Entity -> ABOX_Pancreas-Cell',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> aqueous solution': ('Hydrophilic-Face -> Aqueous-Solution',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'membrane protein -> phospholipid bilayer': ('Membrane-Protein -> Phospholipid-Bilayer',\n",
       "  ['is-inside', 'is-at', 'is-outside']),\n",
       " 'object -> biologist': ('ABOX_Spatial-Entity -> ABOX_Biologist',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> liver cell': ('ABOX_Spatial-Entity -> ABOX_Liver-Cell',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> trachea': ('ABOX_Spatial-Entity -> ABOX_Trachea',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> cytosol': ('ABOX_Hydrophilic-Face -> ABOX_Cytosol',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophobic face -> hydrophobic core': ('ABOX_Hydrophobic-Face -> ABOX_Hydrophobic-Core',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'hydrophilic head -> hydrophilic face': ('ABOX_Hydrophilic-Head -> ABOX_Hydrophilic-Face',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'object -> rotor': ('ABOX_Spatial-Entity -> ABOX_Rotor',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'polypeptide -> cytoplasm': ('ABOX_Polypeptide -> ABOX_Cytoplasm',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'cytosol -> er membrane': ('ABOX_Cytosol -> ABOX_ER-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic end -> inside face': ('ABOX_Hydrophilic-End -> ABOX_Inside-Face',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'hydrophilic face -> membrane': ('ABOX_Hydrophilic-Face -> ABOX_Biomembrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> nucleolemma': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> nuclear membrane': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> nuclear envelope': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> perinuclear-envelope': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> karyotheca': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> nuclear-envelope': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> nucleolemma': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> nuclear membrane': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> nuclear envelope': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> perinuclear-envelope': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> karyotheca': ('Hydrophilic-Face -> Nuclear-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> thylakoid membrane': ('ABOX_Hydrophilic-Face -> ABOX_Thylakoid-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic region -> carbonyl group': ('Hydrophilic-Region -> Carbonyl-Group',\n",
       "  ['is-inside', 'is-at']),\n",
       " 'object -> membrane , which be mitochondrion': ('ABOX_Spatial-Entity -> ABOX_Mitochondrion',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> chlorophyll B': ('ABOX_Spatial-Entity -> ABOX_Chlorophyll-B',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> nucleus': ('ABOX_Spatial-Entity -> ABOX_Nucleus',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> arthropo': ('ABOX_Spatial-Entity -> ABOX_Arthropod',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> animal plasma membrane': ('ABOX_Hydrophilic-Face -> ABOX_Animal-Plasma-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> glucose': ('ABOX_Spatial-Entity -> ABOX_Glucose',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> tropical sea': ('ABOX_Spatial-Entity -> ABOX_Tropical-Sea',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> object': ('ABOX_Spatial-Entity -> ABOX_Tangible-Entity',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> organism': ('ABOX_Tangible-Entity -> ABOX_Organism',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> chloroplast membrane': ('ABOX_Hydrophilic-Face -> ABOX_Chloroplast-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> acoela': ('ABOX_Spatial-Entity -> ABOX_Acoela',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> protist': ('ABOX_Spatial-Entity -> ABOX_Protist',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> flatworm': ('ABOX_Spatial-Entity -> ABOX_Flatworm',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> termite': ('ABOX_Spatial-Entity -> ABOX_Termite',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> mitochondrial membrane': ('Hydrophilic-Face -> Mitochondrial-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'biological region -> organism': ('ABOX_Biological-Region -> ABOX_Organism',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'hydrophilic face -> er membrane': ('ABOX_Hydrophilic-Face -> ABOX_ER-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'outside face of biomembrane -> mitochondrial membrane': ('Hydrophilic-Face -> Mitochondrial-Membrane',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'object -> sunlight': ('ABOX_Spatial-Entity -> ABOX_Sunlight',\n",
       "  ['is-inside', 'is-outside']),\n",
       " 'epinephrine -> extracellular side': ('ABOX_Epinephrine -> ABOX_Extracellular-Side',\n",
       "  ['is-at', 'is-outside']),\n",
       " 'aqueous solution -> hydrophilic head': ('ABOX_Aqueous-Solution -> ABOX_Hydrophilic-Head',\n",
       "  ['abuts', 'is-outside'])}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_relation_mapping = {}\n",
    "for relation in rdb:\n",
    "    if relation == \"no-relation\":\n",
    "        continue\n",
    "    for tp in rdb[relation]:\n",
    "        cp = rdb[relation][tp][\"concept_pair\"]\n",
    "        if tp in term_relation_mapping:\n",
    "            term_relation_mapping[tp][1].append(relation)\n",
    "        else:\n",
    "            term_relation_mapping[tp] = (cp, [relation])\n",
    "multi_label = {tp:r for tp, r in term_relation_mapping.items() if len(r[1]) > 1}\n",
    "print(len(multi_label))\n",
    "#print(np.unique(list(multi_label.values()), return_counts=True))\n",
    "multi_label"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sanity Check Relation Extraction \n",
    "\n",
    "- How many word-pairs match the text on regex, but we don't match in the pipeline?\n",
    "- How many word-pairs don't match sentences? Do these seem reasonable?\n",
    "- Word-pairs with too many matches?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_sentences = read_spacy_docs(\"../data/preprocessed_data/Biology_2e_sentences_spacy\", nlp)\n",
    "bio_sentences += read_spacy_docs(\"../data/preprocessed_data/Life_Biology_sentences_spacy\", nlp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no-relation\n",
      "subclass-of\n",
      "Processing term pair 0/19180\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-fa47d0de780d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     39\u001b[0m             \u001b[0mfound_sentences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0msentence\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mbio_sentences\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m                 \u001b[0msentence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msentence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mterms\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msentence\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m                     \u001b[0mfound_term1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "found_concepts = set()\n",
    "df = {\"sentences\": [], \n",
    "      \"relation\": [], \n",
    "      \"term-pair\": [], \n",
    "      \"concept-pair\": [], \n",
    "      \"tagged\": [], \n",
    "      \"synonym-tagged\": [], \n",
    "      \"count_sentences\": [], \n",
    "      \"textbook_match\": [],\n",
    "      \"term1_found\": [],\n",
    "      \"term2_found\": []} \n",
    "\n",
    "for relation in rdb:\n",
    "    print(relation)\n",
    "    if relation == \"no-relation\":\n",
    "        continue\n",
    "    for i, term_pair in enumerate(rdb[relation]):\n",
    "        if i % 100 == 0:\n",
    "            print(f\"Processing term pair {i}/{len(rdb[relation])}\")\n",
    "        \n",
    "        sentences = rdb[relation][term_pair][\"sentences\"]\n",
    "        count_sentences = len(sentences)\n",
    "        df[\"relation\"].append(relation)\n",
    "        df[\"term-pair\"].append(term_pair)\n",
    "        concept_pair = rdb[relation][term_pair][\"concept_pair\"]\n",
    "        df[\"concept-pair\"].append(concept_pair)\n",
    "        \n",
    "        if count_sentences:\n",
    "            df[\"tagged\"].append(True)\n",
    "            df[\"textbook_match\"].append(True)\n",
    "            df[\"term1_found\"].append(True)\n",
    "            df[\"term2_found\"].append(True)\n",
    "            found_concepts.add(concept_pair)\n",
    "        else:\n",
    "            terms = term_pair.split(\" -> \")\n",
    "            df[\"tagged\"].append(False)\n",
    "            found_term1 = False\n",
    "            found_term2 = False\n",
    "            found_sentences = []\n",
    "            for sentence in bio_sentences:\n",
    "                sentence = str(sentence)\n",
    "                if terms[0] in sentence:\n",
    "                    found_term1 = True\n",
    "                if terms[1] in sentence:\n",
    "                    found_term2 = True\n",
    "                if terms[0] in sentence and terms[1] in sentence:\n",
    "                    found_sentences.append(sentence)\n",
    "            sentences = found_sentences\n",
    "            df[\"term1_found\"].append(found_term1)\n",
    "            df[\"term2_found\"].append(found_term2)\n",
    "            df[\"textbook_match\"].append(len(found_sentences) > 0)\n",
    "                \n",
    "        df[\"count_sentences\"].append(count_sentences)\n",
    "        df[\"sentences\"].append(\"\\n\".join(sentences))\n",
    "                    \n",
    "for cp in df[\"concept-pair\"]:\n",
    "    df[\"synonym-tagged\"].append(cp in found_concepts)\n",
    "\n",
    "df_df = pd.DataFrame(df)\n",
    "df_df.to_excel(f\"{data_dir}/diagnostics/relations_info.xlsx\")\n",
    "df_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_pairs(row):\n",
    "    if row[\"tagged\"]:\n",
    "        group = \"Tagged\"\n",
    "    elif row[\"textbook_match\"]:\n",
    "        group = \"TB_Match_Not_Tagged\"\n",
    "    elif row[\"synonym-tagged\"]:\n",
    "        group = \"Synonym Tagged\"\n",
    "    elif not row[\"term1_found\"] or not row[\"term2_found\"]:\n",
    "        group = \"Missing Term\"\n",
    "    else:\n",
    "        group = \"Terms Not Same Sentence\"\n",
    "    return group\n",
    "\n",
    "df_copy = df_df.copy()\n",
    "df_copy[\"group\"] = df_copy.apply(group_pairs, axis=1)\n",
    "df_copy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_copy.groupby([\"relation\", \"group\"])[\"sentences\"].count().reset_index().rename({\"sentences\": \"count\"}, axis=1).to_csv(f\"{data_dir}/diagnostics/relations_summary.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (conda_pytorch_p36)",
   "language": "python",
   "name": "conda_pytorch_p36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "toc-autonumbering": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
